lahja/endpoint.py

```diff
@@ -61,36 +65,177 @@ def from_name(cls, name: str, base_path: Optional[pathlib.Path]=None) -> 'Connec
+    event: Union[BaseEvent, bytes]
+    config: Optional[BroadcastConfig]
+
+
+class SubscriptionsUpdated(NamedTuple):
+    subscriptions: Set[Type[BaseEvent]]
+    response_expected: bool
+
+
+class SubscriptionsAck:
+    pass
+
+
+Message.register(Broadcast)
+Message.register(SubscriptionsUpdated)
+Message.register(SubscriptionsAck)
+
+
+# mypy doesn't appreciate the ABCMeta trick
+Msg = Union[Broadcast, SubscriptionsUpdated, SubscriptionsAck]
+
+
+class Connection:
+    def __init__(self, reader: StreamReader, writer: StreamWriter) -> None:
+        self.writer = writer
+        self.reader = reader
+        self._drain_lock = asyncio.Lock()
+
+    @staticmethod
+    async def connect_to(path: str) -> 'Connection':
+        reader, writer = await asyncio.open_unix_connection(path)
+        return Connection(reader, writer)
+
+    async def send_message(self, message: Msg) -> None:
+        assert isinstance(message, Message)
+        pickled = pickle.dumps(message)
+        size = len(pickled)
+        self.writer.write(size.to_bytes(4, 'little'))
+        self.writer.write(pickled)
+
+        async with self._drain_lock:
+            # Use a lock to serialize drain() calls. Circumvents this bug:
+            # https://bugs.python.org/issue29930
+            await self.writer.drain()
+
+    async def read_message(self) -> Message:
+        raw_size = await self.reader.readexactly(4)
+        size = int.from_bytes(raw_size, 'little')
+        message = await self.reader.readexactly(size)
+        obj = pickle.loads(message)
+        assert isinstance(obj, Message)
+        return obj
+
+    @property
+    def more_to_read(self) -> bool:
+        return not self.reader.at_eof()
 
 
-class ProxyEndpointConnector(BaseProxy):  # type: ignore # TypeShed is missing BaseProxy
+class InboundConnection:
     """
-    Proxy that connects to an :class:`~lahja.endpoint.EndpointConnector`
+    An Endpoint might have several `InboundConnection`s, each of them represents an
+    Endpoint which has connected to the given Endpoint and is attempting to send it
+    messages.
     """
+    def __init__(self, conn: Connection, new_msg_func: Callable[[Broadcast], None]) -> None:
+        self.conn = conn
+        self.new_msg_func = new_msg_func
+
+        self.logger = logging.getLogger('lahja.endpoint.InboundConnection')
+
+        self._notify_lock = asyncio.Lock()
+        self._received_response = asyncio.Condition()
+
+    async def run(self) -> None:
+        try:
+            while self.conn.more_to_read:
+                message = await self.conn.read_message()
+
+                if isinstance(message, Broadcast):
+                    self.new_msg_func(message)
+                elif isinstance(message, SubscriptionsAck):
+                    async with self._received_response:
+                        self._received_response.notify_all()
+                else:
+                    self.logger.error(f'received unexpected message: {message}')
```

Comment:
Might be good to include the `type(message)` here since it may not be clear what `message` is when represented as a string.